<a href="https://colab.research.google.com/github/mar-valeria/Aprendizaje-autom-tico/blob/main/Mi-segundo-modelo-Plantilla/01-Mi-segundo-modelo-Plantilla.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/andres-merino/AprendizajeAutomaticoInicial-05-N0105/blob/main/2-Ejercicios/01-Mi-segundo-modelo-Plantilla.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<table style="border: none; border-collapse: collapse;">
    <tr>
        <td style="width: 20%; vertical-align: middle; padding-right: 10px;">
            <img src="https://i.imgur.com/nt7hloA.png" width="100">
        </td>
        <td style="width: 2px; text-align: center;">
            <font color="#0030A1" size="7">|</font><br>
            <font color="#0030A1" size="7">|</font>
        </td>
        <td>
            <p style="font-variant: small-caps;"><font color="#0030A1" size="5">
                <b>Facultad de Ciencias Exactas, Naturales y Ambientales</b>
            </font> </p>
            <p style="font-variant: small-caps;"><font color="#0030A1" size="4">
                Aprendizaje Automático Inicial &bull; Mi segundo modelo
            </font></p>
            <p style="font-style: oblique;"><font color="#0030A1" size="3">
                Nombre y Apellido &bull; Fecha
            </font></p>
        </td>  
    </tr>
</table>

---
## <font color='264CC7'> Introducción </font>

Este cuaderno está diseñado practicar en el entrenamiento y evaluación de un modelo de regresión lineal utilizando datos reales. Los objetivos principales son:

- Entrenar un modelo de regresión lineal.
- Evaluar el ajuste del modelo.


Del conjunto de datos disponible [aquí](https://raw.githubusercontent.com/mGalarnyk/Tutorial_Data/master/King_County/kingCountyHouseData.csv), se desea predecir la variable `price` en función de las variables `bedrooms`, `bathrooms`, `sqft_living`, `sqft_lot`, `floors`.

In [325]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, MinMaxScaler, KBinsDiscretizer, OneHotEncoder, RobustScaler, Normalizer, QuantileTransformer, PowerTransformer
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

---
## <font color='264CC7'> Lectura de Datos </font>


Realiza la lectura de los datos y presenta las primeras filas del conjunto de datos.

In [308]:
# Cargar el conjunto de datos
url = 'https://raw.githubusercontent.com/mGalarnyk/Tutorial_Data/master/King_County/kingCountyHouseData.csv'
df = pd.read_csv(url)
# Seleccionar las columnas de interés
columns = ['bedrooms','bathrooms','sqft_living','sqft_lot','floors','price']
df = df.loc[:, columns]
df.head(5)

bedrooms  bathrooms  sqft_living  sqft_lot  floors     price
0         3       1.00         1180      5650     1.0  221900.0
1         3       2.25         2570      7242     2.0  538000.0
2         2       1.00          770     10000     1.0  180000.0
3         4       3.00         1960      5000     1.0  604000.0
4         3       2.00         1680      8080     1.0  510000.0

---
## <font color='264CC7'> Preprocesamiento de datos </font>

Realiza un escalado de los datos, prueba diferentes técnicas y selecciona la que mejor se ajuste a los datos.

In [327]:
# Normalización por el máximo
x_max = df.max()
X_max_normalized = df / x_max
print("Normalización por el máximo:\n")
display(X_max_normalized.head())

# Normalización Min-Max
min_max_scaler = MinMaxScaler()
df_min_max_normalized = pd.DataFrame(min_max_scaler.fit_transform(df), columns=df.columns)
print("Normalización Min-Max:\n")
display(df_min_max_normalized.head())

# Estandarización
standard_scaler = StandardScaler()
X_standardized = pd.DataFrame(standard_scaler.fit_transform(X), columns=X.columns)
print("Estandarización:\n")
display(X_standardized.head())

# RobustScaler
robust_scaler = RobustScaler()
df_robust_scaled = pd.DataFrame(robust_scaler.fit_transform(df), columns=df.columns)
print("Escalado RobustScaler:\n")
display(df_robust_scaled.head())

# Normalizer
normalizer = Normalizer()
df_normalized = pd.DataFrame(normalizer.fit_transform(df), columns=df.columns)
print("Escalado Normalizer:\n")
display(df_normalized.head())

# QuantileTransformer
quantile_scaler = QuantileTransformer(output_distribution='uniform', random_state=0)
df_quantile_scaled = pd.DataFrame(quantile_scaler.fit_transform(df), columns=df.columns)
print("Escalado QuantileTransformer:\n")
display(df_quantile_scaled.head())

# PowerTransformer
power_scaler = PowerTransformer(method='yeo-johnson')
df_power_scaled = pd.DataFrame(power_scaler.fit_transform(df), columns=df.columns)
print("Escalado PowerTransformer:\n")
display(df_power_scaled.head())

Normalización por el máximo:



bedrooms  bathrooms  sqft_living  sqft_lot    floors     price
0  0.090909    0.12500     0.087149  0.003421  0.285714  0.028818
1  0.090909    0.28125     0.189808  0.004385  0.571429  0.069870
2  0.060606    0.12500     0.056869  0.006056  0.285714  0.023377
3  0.121212    0.37500     0.144756  0.003028  0.285714  0.078442
4  0.090909    0.25000     0.124077  0.004893  0.285714  0.066234

Normalización Min-Max:



bedrooms  bathrooms  sqft_living  sqft_lot  floors     price
0  0.090909    0.12500     0.067170  0.003108     0.0  0.019266
1  0.090909    0.28125     0.172075  0.004072     0.4  0.060721
2  0.060606    0.12500     0.036226  0.005743     0.0  0.013770
3  0.121212    0.37500     0.126038  0.002714     0.0  0.069377
4  0.090909    0.25000     0.104906  0.004579     0.0  0.057049

Estandarización:



bedrooms  bathrooms  sqft_living  sqft_lot    floors
0 -0.398737  -1.447464    -0.979835 -0.228321 -0.915427
1 -0.398737   0.175607     0.533634 -0.189885  0.936506
2 -1.473959  -1.447464    -1.426254 -0.123298 -0.915427
3  0.676485   1.149449    -0.130550 -0.244014 -0.915427
4 -0.398737  -0.149007    -0.435422 -0.169653 -0.915427

Escalado RobustScaler:



bedrooms  bathrooms  sqft_living  sqft_lot  floors     price
0       0.0  -1.666667    -0.650045 -0.348442    -0.5 -0.706083
1       0.0   0.000000     0.587711 -0.066572     0.5  0.272404
2      -1.0  -1.666667    -1.015138  0.421742    -0.5 -0.835784
3       1.0   1.000000     0.044524 -0.463527    -0.5  0.476706
4       0.0  -0.333333    -0.204809  0.081799    -0.5  0.185730

Escalado Normalizer:



bedrooms  bathrooms  sqft_living  sqft_lot    floors     price
0  0.000014   0.000005     0.005316  0.025453  0.000005  0.999662
1  0.000006   0.000004     0.004776  0.013460  0.000004  0.999898
2  0.000011   0.000006     0.004271  0.055470  0.000006  0.998451
3  0.000007   0.000005     0.003245  0.008278  0.000002  0.999960
4  0.000006   0.000004     0.003294  0.015841  0.000002  0.999869

Escalado QuantileTransformer:



bedrooms  bathrooms  sqft_living  sqft_lot    floors     price
0  0.358859   0.090090     0.129129  0.301301  0.000000  0.064524
1  0.358859   0.529029     0.754254  0.452221  0.773273  0.630205
2  0.070070   0.090090     0.016517  0.713213  0.000000  0.022022
3  0.748749   0.895896     0.516517  0.238238  0.000000  0.715215
4  0.358859   0.435936     0.380380  0.550012  0.000000  0.591592

Escalado PowerTransformer:



bedrooms  bathrooms  sqft_living  sqft_lot    floors     price
0 -0.352407  -1.571853    -1.121655 -0.331682 -0.975610 -1.487563
1 -0.352407   0.250882     0.706456 -0.032289  1.029432  0.341055
2 -1.605435  -1.571853    -2.107060  0.336051 -0.975610 -1.977885
3  0.722633   1.136536     0.065492 -0.484436 -0.975610  0.553191
4 -0.352407  -0.072697    -0.296964  0.095276 -0.975610  0.241116

---
## <font color='264CC7'> Características y etiquetas </font>

Separa las características y las etiquetas y mostrar la dimensión de cada elemento.

In [318]:
# Dividir el conjunto de datos en características y etiquetas
features = ['bedrooms','bathrooms','sqft_living','sqft_lot','floors']
X = df.loc[:, features]
y = df.loc[:, ['price']]

---
## <font color='264CC7'> Conjunto de entrenamiento y prueba </font>

Divide los datos en dos conjuntos: uno para entrenamiento y otro para prueba y visualiza la dimensión de cada conjunto. Usar `random_state` igual al últmo número de tu cédula.

In [319]:
# Dividir datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

Revisa la distribución de las etiquetas en los conjuntos de entrenamiento y prueba.

In [320]:
print('Dimensiones de los conjuntos de datos:')
print('X_train:', X_train.shape)
print('y_train:', y_train.shape)
print('X_test:', X_test.shape)
print('y_test:', y_test.shape)

Dimensiones de los conjuntos de datos:
X_train: (17290, 5)
y_train: (17290, 1)
X_test: (4323, 5)
y_test: (4323, 1)


---
## <font color='264CC7'> Entrenamiento </font>

Define un modelo de regresión lineal y entrenalo con los datos de entrenamiento.

In [321]:
# Crear el modelo y ajustarlo
modelo = LinearRegression()
modelo.fit(X_train, y_train)

# Visualizar los coeficientes (en escala normalizada)
print('Ecuación del modelo:')
for feature, coef in zip(features, modelo.coef_[0]):
    print(f'{feature}: {coef}')

print('Intercepto:', modelo.intercept_[0])

Ecuación del modelo:
bedrooms: -58230.08137313417
bathrooms: 4992.331204713424
sqft_living: 318.0860369321322
sqft_lot: -0.37174595639226027
floors: 466.5518789340243
Intercepto: 71231.97246657778


---
## <font color='264CC7'> Evaluación </font>

Realiza la predicciones con el conjunto de prueba y muestra los valores en una tabla.

In [322]:
# Predicciones del modelo
y_pred = modelo.predict(X_test)

# Comparar las predicciones con los datos de prueba
df_predicciones = pd.DataFrame({
    'Actual': y_test.values.flatten(),
    'Predicción': y_pred.flatten()
})
display(df_predicciones)

Actual    Predicción
0      297000.0  4.179096e+05
1     1578000.0  1.321662e+06
2      562100.0  4.156510e+05
3      631500.0  3.186865e+05
4      780000.0  8.596693e+05
...         ...           ...
4318   844000.0  8.132561e+05
4319   335500.0  3.033783e+05
4320   369950.0  3.411705e+05
4321   300000.0  2.026461e+05
4322   575950.0  4.254412e+05

[4323 rows x 2 columns]

Calcula el error cuadrático medio.

In [323]:
# Métricas de evaluación
mse_scaled = mean_squared_error(y_test, y_pred_scaled)
print('Error cuadrático medio:', mse_scaled)

Error cuadrático medio: 59742013410.60567


Escribe el error cuadrático medio para cada tipo de escalado:

- Escalado MaxAbsScaler: 59742013410.60567
- Escalado MinMaxScaler: 59742013410.60567
- Escalado RobustScaler: 59742013410.60567
- Escalado StandardScaler: 59742013410.60567
- Escalado Normalizer: 59742013410.60567
- Escalado QuantileTransformer: 59742013410.60567
- Escalado PowerTransformer: 59742013410.60567

Escribe un párrafo sobre qué hace cada escalado no visto en clase:

- **RobustScaler:** Utiliza la mediana y el rango intercuartílico (IQR) en lugar de la media y la desviación estándar. Es ideal cuando hay muchos outliers, ya que no se ve afectado por valores extremos.
- **Normalizer:** Normaliza cada FILA para que tenga norma 1. No escala columnas, sino vectores completos. Es útil en modelos basados en similitud como KNN o clustering sobre ángulos (cosine similarity).
- **QuantileTransformer:** Convierte la distribución original en una distribución uniforme o normal. Garantiza que todos los valores se repartan homogéneamente entre 0 y 1. Es útil cuando los datos no son gaussianos.
- **PowerTransformer:** Aplica transformaciones como Yeo-Johnson o Box-Cox para estabilizar la varianza y hacer la distribución más cercana a una normal. Ayuda cuando hay fuerte asimetría en los datos.